google/cloud/spanner/timestamp.cc

```diff
@@ -52,347 +47,66 @@ Status NegativeOverflow(std::string const& type) {
-      return NegativeOverflow(kDestType);
-    }
-  }
-  return FromCounts(sec + utc_offset_secs, nanos);
-}
-
-// TODO(#145): Reconcile this implementation with FormatRfc3339() in
-// google/cloud/internal/format_time_point.h in google-cloud-cpp.
-std::string Timestamp::ToRFC3339() const {
-  std::ostringstream output;
-
-  // Spanner always uses "Z" but we leave support for a non-zero UTC offset
-  // for later refactoring of this code to more general scenarios.
-  constexpr std::int64_t kUtcOffsetSecs = 0;
-
-  // Note: FormatTime(ZTime()) can only do the right thing when the requested
-  // time is within the range of a std::tm (to wit, the "int tm_year" field).
-  output << internal::FormatTime(ZTime(sec_ + kUtcOffsetSecs));
-
-  if (auto ss = nsec_) {
-    int width = 0;
-    auto scale = kNanosPerSecond;
-    while (scale != 1) {
-      scale /= 10;
-      width += 1;
-    }
-    while (ss % 10 == 0) {
-      ss /= 10;
-      width -= 1;
-    }
-    output << '.' << std::setfill('0') << std::setw(width) << ss;
-  }
-
-  if (kUtcOffsetSecs != 0) {
-    // Format colon-separated hours, minutes, but not (yet) seconds.
-    output << std::setfill('0') << std::internal << std::showpos
-           << std::setw(1 + 2) << (kUtcOffsetSecs / kSecsPerHour) << ':'
-           << std::noshowpos << std::setw(2)
-           << std::abs(kUtcOffsetSecs) % kSecsPerHour / kSecsPerMinute;
-  } else {
-    output << 'Z';
-  }
-  return output.str();
-}
-
-Timestamp Timestamp::FromProto(protobuf::Timestamp const& proto) {
-  auto ts = FromCounts(proto.seconds(), proto.nanos());
-  if (!ts) {
-    // If the proto cannot be normalized (proto.nanos() would need to be
-    // outside its documented [0..999999999] range and have the same sign
-    // as proto.seconds()), then we saturate.
-    return proto.seconds() >= 0
-               ? Timestamp{std::numeric_limits<std::int64_t>::max(),
-                           kNanosPerSecond - 1}
-               : Timestamp{std::numeric_limits<std::int64_t>::min(), 0};
-  }
-  return *ts;
-}
-
-protobuf::Timestamp Timestamp::ToProto() const {
-  // May produce a protobuf::Timestamp outside the documented range of
-  // 0001-01-01T00:00:00Z to 9999-12-31T23:59:59.999999999Z inclusive,
-  // but so be it.
-  protobuf::Timestamp proto;
-  proto.set_seconds(sec_);
-  proto.set_nanos(nsec_);
-  return proto;
+StatusOr<std::intmax_t> Timestamp::ToRatio(std::intmax_t min, std::intmax_t max,
+                                           std::intmax_t num,
+                                           std::intmax_t den) const {
+  constexpr auto kDestType = "std::chrono::time_point";
+  auto const period = absl::Seconds(num) / den;
+  auto const duration = absl::Floor(t_ - absl::UnixEpoch(), period);
+  if (duration > max * period) return PositiveOverflow(kDestType);
+  if (duration < min * period) return NegativeOverflow(kDestType);
+  return duration / period;
```

Comment:
Let me try to clarify a bit more, then if it doesn't help, let's jump to a GVC. Of course, maybe I'm confused and the existing code is just wrong. In that case, I'd be happy to change it to something correct :-)

Assumptions:
* `Timestamp` has a range of 10k years.
* `absl::Time`'s range is much wider, in the billions of years.
* The ratio `num/den` will never be finer than "nanoseconds".
* The caller cannot have a "count" (i.e., a `Duration::rep`) wider than `std::int64_t`.

Given the above assumptions, I believe the only overflow we're concerned about is potentially returning a `count` that is too big for the caller's `Duration::rep`. `min` and `max` tell us the range of the caller's `Duration::rep` so we can do the overflow check in this function rather than having to do it in the calling `ConvertTo` template.

The only expression that could potentially overflow is `duration / period`. As you know, if this expression overflows, it just saturates at int64 min/max, so we wouldn't know if it did overflow. However, `absl::Duration` has a wider range than int64, so we can do the overflow check in that domain instead. 

Logically, the overflow condition I _want_ to check is (for the positive overflow case) `if (duration / period > max)`. However, that computation is done in int64 space, and wouldn't be able to detect all overflows. So an equivalent expression is (multiplying both sides by period) `if (duration > max * period)`, which performs the arithmetic in the `absl::Duration` space, which is wide enough to catch the overflow.

Reply:
I don't think the `absl::Duration` space is wide enough.  `max * period` could exceed its range when `period > Seconds(1)`.  [Ditto for `min` but let's stick with the positive case.]

Still, we do get reliable semantics from `absl::Duration` ... the result of the overflow multiplication will be `absl::InfiniteDuration()`.  Nothing is greater than that, so the question becomes: if `max * period` overflows, is `duration / period <= max`?  That's what I hadn't convinced myself of yesterday, but, given the limited magnitude of `duration`, I'm now convinced of it.

The "good" thing about the `absl::IDivDuration()` incantation is that it tells us directly whether `duration / period` is representable in 64 bits, and, if so, the true quotient.  It is also nice to do one division instead of two divisions and two multiplications.

So, yes, let's GVC and come to a conclusion.

Reply:
I'm happy that the code as-is is producing the desired results.  We had slight dissonance on the obviousness of "equivalent expression", but so be it.